File: backend/provider-limits/codex.ts
// ---- Codex provider adapter (faithful port of the legacy provider-limits Codex adapter) ----

import { spawn, execSync } from "node:child_process";
import { createReadStream, existsSync, readdirSync, statSync } from "node:fs";
import { createInterface } from "node:readline";
import { homedir } from "node:os";
import { join } from "node:path";

import type {
  ProviderLimitSnapshot,
  ProviderLimitWindow,
  ProviderCredits,
} from "./models.js";

// ---------------------------------------------------------------------------
// Codex app-server RPC response shapes (camelCase from the protocol)
// ---------------------------------------------------------------------------

interface RateLimitWindow {
  usedPercent: number;
  windowDurationMins?: number | null;
  resetsAt?: number | null;
}

interface CreditsSnapshot {
  hasCredits: boolean;
  unlimited: boolean;
  balance?: string | null;
}

interface RateLimitSnapshot {
  primary?: RateLimitWindow | null;
  secondary?: RateLimitWindow | null;
  credits?: CreditsSnapshot | null;
  planType?: string | null;
}

interface RateLimitsResult {
  rateLimits: RateLimitSnapshot;
}

interface AccountInfo {
  type: string;
  email?: string | null;
  planType?: string | null;
}

interface AccountResult {
  account: AccountInfo;
}

// ---------------------------------------------------------------------------
// Session log fallback shapes (snake_case from JSONL files)
// ---------------------------------------------------------------------------

interface SessionLogWindow {
  used_percent: number;
  window_minutes?: number | null;
  resets_at?: number | null;
}

interface SessionLogCredits {
  has_credits?: boolean | null;
  unlimited?: boolean | null;
  balance?: string | null;
}

interface SessionLogRateLimits {
  primary?: SessionLogWindow | null;
  secondary?: SessionLogWindow | null;
  credits?: SessionLogCredits | null;
  plan_type?: string | null;
}

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

const STALE_AFTER_SECONDS = 900;
const PROBE_TIMEOUT_SECONDS = 15;

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

export async function fetch_codex_limits(): Promise<ProviderLimitSnapshot> {
  try {
    return await probe_app_server();
  } catch (rpc_err) {
    console.error(`[provider_limits::codex] app-server probe failed: ${rpc_err}`);
    try {
      return await fallback_session_logs();
    } catch (log_err) {
      console.error(`[provider_limits::codex] session log fallback failed: ${log_err}`);
      return make_error_snapshot(`app-server: ${rpc_err}; logs: ${log_err}`);
    }
  }
}

// ---------------------------------------------------------------------------
// App-server probe
// ---------------------------------------------------------------------------

async function probe_app_server(): Promise<ProviderLimitSnapshot> {
  const codex_bin = find_codex_binary();
  if (codex_bin === null) {
    throw new Error("codex binary not found -- is Codex CLI installed?");
  }

  const child = spawn(codex_bin, ["app-server"], {
    stdio: ["pipe", "pipe", "ignore"],
  });

  return new Promise<ProviderLimitSnapshot>((resolve, reject) => {
    const timeout = setTimeout(() => {
      child.kill();
      reject(new Error("codex app-server probe timed out"));
    }, PROBE_TIMEOUT_SECONDS * 1000);

    run_rpc_exchange(child)
      .then((snapshot) => {
        clearTimeout(timeout);
        child.kill();
        resolve(snapshot);
      })
      .catch((err) => {
        clearTimeout(timeout);
        child.kill();
        reject(err);
      });
  });
}

async function run_rpc_exchange(
  child: ReturnType<typeof spawn>,
): Promise<ProviderLimitSnapshot> {
  const stdin = child.stdin!;
  const rl = createInterface({ input: child.stdout! });

  const pending_responses = new Map<
    number,
    { resolve: (v: unknown) => void; reject: (e: Error) => void }
  >();

  rl.on("line", (raw_line: string) => {
    const trimmed = raw_line.trim();
    if (trimmed === "") return;

    let parsed: Record<string, unknown>;
    try {
      parsed = JSON.parse(trimmed);
    } catch {
      return; // skip malformed lines
    }

    const id = typeof parsed.id === "number" ? parsed.id : null;
    if (id === null) return; // notification, skip

    const entry = pending_responses.get(id);
    if (!entry) return;
    pending_responses.delete(id);

    if (parsed.error) {
      entry.reject(new Error(`codex RPC error: ${JSON.stringify(parsed.error)}`));
      return;
    }
    if (parsed.result !== undefined) {
      entry.resolve(parsed.result);
      return;
    }
    entry.reject(new Error("codex RPC response missing 'result' field"));
  });

  function send_and_wait(msg: Record<string, unknown>, id: number): Promise<unknown> {
    return new Promise((resolve, reject) => {
      pending_responses.set(id, { resolve, reject });
      stdin.write(JSON.stringify(msg) + "\n");
    });
  }

  // 1. Initialize
  await send_and_wait(
    {
      method: "initialize",
      id: 1,
      params: {
        clientInfo: { name: "ariadne", title: null, version: "0.1.0" },
      },
    },
    1,
  );

  // 2. account/rateLimits/read
  const limits_resp = await send_and_wait(
    { method: "account/rateLimits/read", id: 2 },
    2,
  );

  // 3. account/read
  const account_resp = await send_and_wait(
    { method: "account/read", id: 3, params: { refreshToken: false } },
    3,
  );

  rl.close();

  // Parse rate limits
  const limits_result = limits_resp as RateLimitsResult;
  if (!limits_result?.rateLimits) {
    throw new Error("Failed to parse rateLimits response: missing rateLimits field");
  }

  // Parse account (optional -- don't fail if it doesn't parse)
  let account_result: AccountResult | null = null;
  try {
    const ar = account_resp as AccountResult;
    if (ar?.account) {
      account_result = ar;
    }
  } catch {
    // ignore
  }

  const now = new Date();
  const snap = limits_result.rateLimits;

  const windows: ProviderLimitWindow[] = [];
  if (snap.primary) {
    windows.push(normalize_window("primary", "Primary (5h)", snap.primary));
  }
  if (snap.secondary) {
    windows.push(normalize_window("secondary", "Secondary (7d)", snap.secondary));
  }

  const credits: ProviderCredits | null = snap.credits
    ? {
        has_credits: snap.credits.hasCredits,
        unlimited: snap.credits.unlimited,
        balance: snap.credits.balance ?? null,
      }
    : null;

  const account_label = account_result?.account?.email ?? null;
  const plan_type =
    snap.planType ?? account_result?.account?.planType ?? null;

  return {
    provider_id: "codex",
    provider_label: "Codex",
    account_label,
    plan_type,
    source: "codex-app-server",
    source_confidence: "high",
    status: "fresh",
    fetched_at: now.toISOString(),
    stale_after_seconds: STALE_AFTER_SECONDS,
    windows,
    credits,
    error_message: null,
  };
}

function normalize_window(
  id: string,
  label: string,
  w: RateLimitWindow,
): ProviderLimitWindow {
  let resets_at_iso: string | null = null;
  if (w.resetsAt != null) {
    try {
      resets_at_iso = epoch_to_rfc3339(w.resetsAt);
    } catch {
      resets_at_iso = null;
    }
  }

  return {
    id,
    label,
    used_percent: w.usedPercent,
    remaining_percent: 100.0 - w.usedPercent,
    window_minutes: w.windowDurationMins ?? null,
    resets_at: resets_at_iso,
  };
}

// ---------------------------------------------------------------------------
// Session log fallback
// ---------------------------------------------------------------------------

export async function fallback_session_logs(): Promise<ProviderLimitSnapshot> {
  const override_home = process.env.ARIADNE_CODEX_HOME;
  const codex_home = override_home ?? join(homedir(), ".codex");
  const sessions_dir = join(codex_home, "sessions");

  if (!existsSync(sessions_dir)) {
    throw new Error("~/.codex/sessions does not exist");
  }

  // Collect all .jsonl files recursively
  const jsonl_files: string[] = [];
  collect_jsonl_files(sessions_dir, jsonl_files);

  if (jsonl_files.length === 0) {
    throw new Error("No session log files found");
  }

  // Sort by modified time descending (newest first)
  jsonl_files.sort((a, b) => {
    const ma = safe_mtime(a);
    const mb = safe_mtime(b);
    return mb - ma;
  });

  // Scan up to 20 most recent files
  for (const file_path of jsonl_files.slice(0, 20)) {
    const snapshot = await try_extract_rate_limits(file_path);
    if (snapshot !== null) {
      return snapshot;
    }
  }

  throw new Error("No rate_limits found in recent session logs");
}

function safe_mtime(file_path: string): number {
  try {
    return statSync(file_path).mtimeMs;
  } catch {
    return 0;
  }
}

function collect_jsonl_files(dir: string, out: string[]): void {
  let names: string[];
  try {
    names = readdirSync(dir);
  } catch {
    return;
  }

  for (const name of names) {
    const full_path = join(dir, name);
    try {
      const stat = statSync(full_path);
      if (stat.isDirectory()) {
        collect_jsonl_files(full_path, out);
      } else if (name.endsWith(".jsonl")) {
        out.push(full_path);
      }
    } catch {
      // skip inaccessible entries
    }
  }
}

async function try_extract_rate_limits(
  file_path: string,
): Promise<ProviderLimitSnapshot | null> {
  return new Promise((resolve, reject) => {
    let last_rate_limits: SessionLogRateLimits | null = null;
    let last_timestamp: string | null = null;

    const stream = createReadStream(file_path, { encoding: "utf8" });
    const rl = createInterface({ input: stream, crlfDelay: Infinity });

    rl.on("line", (line: string) => {
      const trimmed = line.trim();
      if (trimmed === "" || !trimmed.includes("rate_limits")) {
        return;
      }

      let entry: Record<string, unknown>;
      try {
        entry = JSON.parse(trimmed);
      } catch {
        return;
      }

      const payload = entry.payload as Record<string, unknown> | undefined;
      if (!payload) return;

      const rl_val = payload.rate_limits;
      if (!rl_val || typeof rl_val !== "object") return;

      // Treat as SessionLogRateLimits
      const rl_parsed = rl_val as SessionLogRateLimits;
      last_rate_limits = rl_parsed;
      last_timestamp =
        typeof entry.timestamp === "string" ? entry.timestamp : null;
    });

    rl.on("close", () => {
      if (last_rate_limits === null) {
        resolve(null);
        return;
      }

      const rl = last_rate_limits;
      const now = new Date();

      // Determine age and status
      let status: "fresh" | "stale" = "stale";
      let fetched_at: string = now.toISOString();

      if (last_timestamp !== null) {
        const dt = new Date(last_timestamp);
        if (!isNaN(dt.getTime())) {
          const age_secs = Math.abs(
            Math.floor((now.getTime() - dt.getTime()) / 1000),
          );
          status = age_secs <= STALE_AFTER_SECONDS ? "fresh" : "stale";
          fetched_at = last_timestamp;
        }
      }

      const windows: ProviderLimitWindow[] = [];
      if (rl.primary) {
        windows.push(
          normalize_log_window("primary", "Primary (5h)", rl.primary),
        );
      }
      if (rl.secondary) {
        windows.push(
          normalize_log_window("secondary", "Secondary (7d)", rl.secondary),
        );
      }

      const credits: ProviderCredits | null = rl.credits
        ? {
            has_credits: rl.credits.has_credits ?? false,
            unlimited: rl.credits.unlimited ?? false,
            balance: rl.credits.balance ?? null,
          }
        : null;

      resolve({
        provider_id: "codex",
        provider_label: "Codex",
        account_label: null,
        plan_type: rl.plan_type ?? null,
        source: "codex-session-log",
        source_confidence: "medium",
        status,
        fetched_at,
        stale_after_seconds: STALE_AFTER_SECONDS,
        windows,
        credits,
        error_message: null,
      });
    });

    rl.on("error", (err: Error) => {
      reject(new Error(`Failed to read ${file_path}: ${err.message}`));
    });

    stream.on("error", (err: Error) => {
      reject(new Error(`Failed to open ${file_path}: ${err.message}`));
    });
  });
}

function normalize_log_window(
  id: string,
  label: string,
  w: SessionLogWindow,
): ProviderLimitWindow {
  let resets_at_iso: string | null = null;
  if (w.resets_at != null) {
    try {
      resets_at_iso = epoch_to_rfc3339(w.resets_at);
    } catch {
      resets_at_iso = null;
    }
  }

  return {
    id,
    label,
    used_percent: w.used_percent,
    remaining_percent: 100.0 - w.used_percent,
    window_minutes: w.window_minutes ?? null,
    resets_at: resets_at_iso,
  };
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

/**
 * Convert a unix epoch (seconds) to an RFC 3339 string matching Rust's
 * chrono::to_rfc3339() output: "2025-03-30T18:00:00+00:00".
 * JavaScript's Date.toISOString() produces "...Z" with milliseconds which
 * differs from the Rust format.
 */
function epoch_to_rfc3339(epoch_secs: number): string {
  const d = new Date(epoch_secs * 1000);
  const iso = d.toISOString(); // e.g. "2025-03-30T18:00:00.000Z"
  // Strip trailing milliseconds + Z, append +00:00
  return iso.replace(/\.(\d{3})Z$/, "+00:00");
}

export function find_codex_binary(): string | null {
  const home = homedir();
  const candidates = [
    join(home, ".bun", "bin", "codex"),
    "/opt/homebrew/bin/codex",
    "/usr/local/bin/codex",
  ];

  for (const candidate of candidates) {
    if (existsSync(candidate)) {
      return candidate;
    }
  }

  // Fall back to PATH lookup via `which`
  try {
    const result = execSync("which codex", { encoding: "utf8", stdio: ["pipe", "pipe", "pipe"] }).trim();
    if (result !== "") {
      return result;
    }
  } catch {
    // which failed
  }

  return null;
}

export function make_error_snapshot(error_message: string): ProviderLimitSnapshot {
  return {
    provider_id: "codex",
    provider_label: "Codex",
    account_label: null,
    plan_type: null,
    source: "none",
    source_confidence: "low",
    status: "error",
    fetched_at: new Date().toISOString(),
    stale_after_seconds: STALE_AFTER_SECONDS,
    windows: [],
    credits: null,
    error_message,
  };
}

File: src-tauri/src/provider_limits/codex.rs
use chrono::{DateTime, TimeZone, Utc};
use serde::Deserialize;
use serde_json::{json, Value};
use std::path::PathBuf;
use tokio::io::{AsyncBufReadExt, AsyncWriteExt, BufReader};
use tokio::process::Command;

use crate::models::provider_limits::{
    ProviderCredits, ProviderLimitSnapshot, ProviderLimitWindow, SnapshotStatus, SourceConfidence,
};

// ---------------------------------------------------------------------------
// Codex app-server RPC response shapes (camelCase from the protocol)
// ---------------------------------------------------------------------------

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
struct RateLimitsResult {
    rate_limits: RateLimitSnapshot,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
struct RateLimitSnapshot {
    primary: Option<RateLimitWindow>,
    secondary: Option<RateLimitWindow>,
    credits: Option<CreditsSnapshot>,
    plan_type: Option<String>,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
struct RateLimitWindow {
    used_percent: f64,
    window_duration_mins: Option<u32>,
    resets_at: Option<i64>,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
struct CreditsSnapshot {
    has_credits: bool,
    unlimited: bool,
    balance: Option<String>,
}

#[derive(Debug, Deserialize)]
struct AccountResult {
    account: AccountInfo,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
struct AccountInfo {
    #[serde(rename = "type")]
    _auth_type: String,
    email: Option<String>,
    plan_type: Option<String>,
}

// ---------------------------------------------------------------------------
// Session log fallback shapes (snake_case from JSONL files)
// ---------------------------------------------------------------------------

#[derive(Debug, Deserialize)]
struct SessionLogRateLimits {
    primary: Option<SessionLogWindow>,
    secondary: Option<SessionLogWindow>,
    credits: Option<SessionLogCredits>,
    plan_type: Option<String>,
}

#[derive(Debug, Deserialize)]
struct SessionLogWindow {
    used_percent: f64,
    window_minutes: Option<u32>,
    resets_at: Option<i64>,
}

#[derive(Debug, Deserialize)]
struct SessionLogCredits {
    has_credits: Option<bool>,
    unlimited: Option<bool>,
    balance: Option<String>,
}

/// Stale-after threshold in seconds (15 minutes).
const STALE_AFTER_SECONDS: u64 = 900;

/// Timeout for the entire app-server probe in seconds.
const PROBE_TIMEOUT_SECONDS: u64 = 15;

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

/// Fetch Codex provider limits. Tries the app-server probe first, falls back
/// to the most recent session log entry with rate_limits.
pub async fn fetch_codex_limits() -> ProviderLimitSnapshot {
    match probe_app_server().await {
        Ok(snapshot) => snapshot,
        Err(rpc_err) => {
            eprintln!("[provider_limits::codex] app-server probe failed: {rpc_err}");
            match fallback_session_logs().await {
                Ok(snapshot) => snapshot,
                Err(log_err) => {
                    eprintln!("[provider_limits::codex] session log fallback failed: {log_err}");
                    make_error_snapshot(format!("app-server: {rpc_err}; logs: {log_err}"))
                }
            }
        }
    }
}

// ---------------------------------------------------------------------------
// App-server probe
// ---------------------------------------------------------------------------

async fn probe_app_server() -> Result<ProviderLimitSnapshot, String> {
    let codex_bin = find_codex_binary()?;

    let mut child = Command::new(&codex_bin)
        .arg("app-server")
        .stdin(std::process::Stdio::piped())
        .stdout(std::process::Stdio::piped())
        .stderr(std::process::Stdio::null())
        .kill_on_drop(true)
        .spawn()
        .map_err(|e| format!("Failed to spawn codex app-server: {e}"))?;

    let stdin = child
        .stdin
        .take()
        .ok_or("Failed to open stdin on codex app-server")?;
    let stdout = child
        .stdout
        .take()
        .ok_or("Failed to open stdout on codex app-server")?;

    let result = tokio::time::timeout(
        std::time::Duration::from_secs(PROBE_TIMEOUT_SECONDS),
        run_rpc_exchange(stdin, stdout),
    )
    .await
    .map_err(|_| "codex app-server probe timed out".to_string())?;

    // Ensure the child is killed regardless of outcome.
    let _ = child.kill().await;

    result
}

async fn run_rpc_exchange(
    mut stdin: tokio::process::ChildStdin,
    stdout: tokio::process::ChildStdout,
) -> Result<ProviderLimitSnapshot, String> {
    let mut reader = BufReader::new(stdout).lines();

    // 1. Initialize
    let init_msg = json!({
        "method": "initialize",
        "id": 1,
        "params": {
            "clientInfo": {
                "name": "ariadne",
                "title": null,
                "version": "0.1.0"
            }
        }
    });
    send_line(&mut stdin, &init_msg).await?;
    read_response(&mut reader, 1).await?; // consume init response

    // 2. account/rateLimits/read
    let limits_msg = json!({
        "method": "account/rateLimits/read",
        "id": 2
    });
    send_line(&mut stdin, &limits_msg).await?;
    let limits_resp = read_response(&mut reader, 2).await?;

    // 3. account/read
    let account_msg = json!({
        "method": "account/read",
        "id": 3,
        "params": { "refreshToken": false }
    });
    send_line(&mut stdin, &account_msg).await?;
    let account_resp = read_response(&mut reader, 3).await?;

    // Parse rate limits
    let limits_result: RateLimitsResult = serde_json::from_value(limits_resp)
        .map_err(|e| format!("Failed to parse rateLimits response: {e}"))?;

    // Parse account (optional — don't fail if this doesn't parse)
    let account_result: Option<AccountResult> =
        serde_json::from_value(account_resp).ok();

    let now = Utc::now();
    let snap = &limits_result.rate_limits;

    let mut windows = Vec::new();
    if let Some(ref primary) = snap.primary {
        windows.push(normalize_window("primary", "Primary (5h)", primary));
    }
    if let Some(ref secondary) = snap.secondary {
        windows.push(normalize_window("secondary", "Secondary (7d)", secondary));
    }

    let credits = snap.credits.as_ref().map(|c| ProviderCredits {
        has_credits: c.has_credits,
        unlimited: c.unlimited,
        balance: c.balance.clone(),
    });

    let account_label = account_result
        .as_ref()
        .and_then(|a| a.account.email.clone());

    let plan_type = snap
        .plan_type
        .clone()
        .or_else(|| account_result.and_then(|a| a.account.plan_type));

    Ok(ProviderLimitSnapshot {
        provider_id: "codex".to_string(),
        provider_label: "Codex".to_string(),
        account_label,
        plan_type,
        source: "codex-app-server".to_string(),
        source_confidence: SourceConfidence::High,
        status: SnapshotStatus::Fresh,
        fetched_at: now.to_rfc3339(),
        stale_after_seconds: STALE_AFTER_SECONDS,
        windows,
        credits,
        error_message: None,
    })
}

async fn send_line(
    stdin: &mut tokio::process::ChildStdin,
    msg: &Value,
) -> Result<(), String> {
    let mut line = serde_json::to_string(msg).map_err(|e| format!("JSON serialize error: {e}"))?;
    line.push('\n');
    stdin
        .write_all(line.as_bytes())
        .await
        .map_err(|e| format!("Failed to write to codex stdin: {e}"))?;
    stdin
        .flush()
        .await
        .map_err(|e| format!("Failed to flush codex stdin: {e}"))?;
    Ok(())
}

async fn read_response(
    reader: &mut tokio::io::Lines<BufReader<tokio::process::ChildStdout>>,
    expected_id: u32,
) -> Result<Value, String> {
    // Read lines until we get a response matching the expected ID.
    // The server may emit notifications between responses.
    loop {
        let line = reader
            .next_line()
            .await
            .map_err(|e| format!("Failed to read from codex stdout: {e}"))?
            .ok_or("codex app-server closed stdout unexpectedly")?;

        let trimmed = line.trim();
        if trimmed.is_empty() {
            continue;
        }

        let parsed: Value = serde_json::from_str(trimmed)
            .map_err(|e| format!("Malformed JSON from codex: {e}"))?;

        // Check if this is a response (has "id" field) matching our expected ID.
        if let Some(id) = parsed.get("id").and_then(|v| v.as_u64()) {
            if id == expected_id as u64 {
                // Check for RPC error
                if let Some(err) = parsed.get("error") {
                    return Err(format!("codex RPC error: {err}"));
                }
                if let Some(result) = parsed.get("result") {
                    return Ok(result.clone());
                }
                return Err("codex RPC response missing 'result' field".to_string());
            }
        }
        // Otherwise it's a notification or a response for a different ID; skip.
    }
}

fn normalize_window(id: &str, label: &str, w: &RateLimitWindow) -> ProviderLimitWindow {
    let resets_at_iso = w.resets_at.and_then(|ts| {
        Utc.timestamp_opt(ts, 0)
            .single()
            .map(|dt| dt.to_rfc3339())
    });

    ProviderLimitWindow {
        id: id.to_string(),
        label: label.to_string(),
        used_percent: Some(w.used_percent),
        remaining_percent: Some(100.0 - w.used_percent),
        window_minutes: w.window_duration_mins,
        resets_at: resets_at_iso,
    }
}

// ---------------------------------------------------------------------------
// Session log fallback
// ---------------------------------------------------------------------------

async fn fallback_session_logs() -> Result<ProviderLimitSnapshot, String> {
    let home = dirs::home_dir().ok_or("Cannot determine home directory")?;
    let sessions_dir = home.join(".codex").join("sessions");

    if !sessions_dir.exists() {
        return Err("~/.codex/sessions does not exist".to_string());
    }

    // Collect all .jsonl files, sorted by modified time descending.
    let mut jsonl_files: Vec<PathBuf> = Vec::new();
    collect_jsonl_files(&sessions_dir, &mut jsonl_files);

    if jsonl_files.is_empty() {
        return Err("No session log files found".to_string());
    }

    // Sort by modified time descending (newest first).
    jsonl_files.sort_by(|a, b| {
        let ma = std::fs::metadata(a)
            .and_then(|m| m.modified())
            .unwrap_or(std::time::UNIX_EPOCH);
        let mb = std::fs::metadata(b)
            .and_then(|m| m.modified())
            .unwrap_or(std::time::UNIX_EPOCH);
        mb.cmp(&ma)
    });

    // Scan recent files (limit to newest 20) for the latest rate_limits entry.
    for path in jsonl_files.iter().take(20) {
        if let Some(snapshot) = try_extract_rate_limits(path)? {
            return Ok(snapshot);
        }
    }

    Err("No rate_limits found in recent session logs".to_string())
}

fn collect_jsonl_files(dir: &std::path::Path, out: &mut Vec<PathBuf>) {
    if let Ok(entries) = std::fs::read_dir(dir) {
        for entry in entries.flatten() {
            let path = entry.path();
            if path.is_dir() {
                collect_jsonl_files(&path, out);
            } else if path.extension().map(|e| e == "jsonl").unwrap_or(false) {
                out.push(path);
            }
        }
    }
}

fn try_extract_rate_limits(
    path: &std::path::Path,
) -> Result<Option<ProviderLimitSnapshot>, String> {
    use std::io::{BufRead, BufReader};

    let file = std::fs::File::open(path)
        .map_err(|e| format!("Failed to open {}: {e}", path.display()))?;
    let reader = BufReader::new(file);

    let mut last_rate_limits: Option<SessionLogRateLimits> = None;
    let mut last_timestamp: Option<String> = None;

    for line in reader.lines() {
        let line = line.map_err(|e| format!("Read error in {}: {e}", path.display()))?;
        let trimmed = line.trim();
        if trimmed.is_empty() || !trimmed.contains("rate_limits") {
            continue;
        }

        if let Ok(entry) = serde_json::from_str::<Value>(trimmed) {
            // Try to extract rate_limits from the payload
            if let Some(payload) = entry.get("payload") {
                if let Some(rl_val) = payload.get("rate_limits") {
                    if let Ok(rl) = serde_json::from_value::<SessionLogRateLimits>(rl_val.clone()) {
                        last_timestamp = entry
                            .get("timestamp")
                            .and_then(|t| t.as_str())
                            .map(String::from);
                        last_rate_limits = Some(rl);
                    }
                }
            }
        }
    }

    let rl = match last_rate_limits {
        Some(rl) => rl,
        None => return Ok(None),
    };

    let now = Utc::now();

    // Determine age — if we have a timestamp from the log, compute staleness.
    let (status, fetched_at) = if let Some(ref ts) = last_timestamp {
        if let Ok(dt) = DateTime::parse_from_rfc3339(ts) {
            let age_secs = now
                .signed_duration_since(dt.with_timezone(&Utc))
                .num_seconds()
                .unsigned_abs();
            let st = if age_secs <= STALE_AFTER_SECONDS {
                SnapshotStatus::Fresh
            } else {
                SnapshotStatus::Stale
            };
            (st, ts.clone())
        } else {
            (SnapshotStatus::Stale, now.to_rfc3339())
        }
    } else {
        (SnapshotStatus::Stale, now.to_rfc3339())
    };

    let mut windows = Vec::new();
    if let Some(ref primary) = rl.primary {
        windows.push(normalize_log_window("primary", "Primary (5h)", primary));
    }
    if let Some(ref secondary) = rl.secondary {
        windows.push(normalize_log_window(
            "secondary",
            "Secondary (7d)",
            secondary,
        ));
    }

    let credits = rl.credits.as_ref().map(|c| ProviderCredits {
        has_credits: c.has_credits.unwrap_or(false),
        unlimited: c.unlimited.unwrap_or(false),
        balance: c.balance.clone(),
    });

    Ok(Some(ProviderLimitSnapshot {
        provider_id: "codex".to_string(),
        provider_label: "Codex".to_string(),
        account_label: None,
        plan_type: rl.plan_type,
        source: "codex-session-log".to_string(),
        source_confidence: SourceConfidence::Medium,
        status,
        fetched_at,
        stale_after_seconds: STALE_AFTER_SECONDS,
        windows,
        credits,
        error_message: None,
    }))
}

fn normalize_log_window(
    id: &str,
    label: &str,
    w: &SessionLogWindow,
) -> ProviderLimitWindow {
    let resets_at_iso = w.resets_at.and_then(|ts| {
        Utc.timestamp_opt(ts, 0)
            .single()
            .map(|dt| dt.to_rfc3339())
    });

    ProviderLimitWindow {
        id: id.to_string(),
        label: label.to_string(),
        used_percent: Some(w.used_percent),
        remaining_percent: Some(100.0 - w.used_percent),
        window_minutes: w.window_minutes,
        resets_at: resets_at_iso,
    }
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

fn find_codex_binary() -> Result<String, String> {
    // Check common locations
    let candidates = [
        // Bun global installs
        dirs::home_dir().map(|h| h.join(".bun/bin/codex")),
        // Homebrew
        Some(PathBuf::from("/opt/homebrew/bin/codex")),
        Some(PathBuf::from("/usr/local/bin/codex")),
    ];

    for candidate in &candidates {
        if let Some(path) = candidate {
            if path.exists() {
                return Ok(path.to_string_lossy().to_string());
            }
        }
    }

    // Fall back to PATH lookup via `which`
    if let Ok(output) = std::process::Command::new("which")
        .arg("codex")
        .output()
    {
        if output.status.success() {
            let path = String::from_utf8_lossy(&output.stdout).trim().to_string();
            if !path.is_empty() {
                return Ok(path);
            }
        }
    }

    Err("codex binary not found — is Codex CLI installed?".to_string())
}

fn make_error_snapshot(error_message: String) -> ProviderLimitSnapshot {
    ProviderLimitSnapshot {
        provider_id: "codex".to_string(),
        provider_label: "Codex".to_string(),
        account_label: None,
        plan_type: None,
        source: "none".to_string(),
        source_confidence: SourceConfidence::Low,
        status: SnapshotStatus::Error,
        fetched_at: Utc::now().to_rfc3339(),
        stale_after_seconds: STALE_AFTER_SECONDS,
        windows: vec![],
        credits: None,
        error_message: Some(error_message),
    }
}
